usr/src/uts/common/fs/zfs/zfs_dir.c

```diff
@@ -609,16 +609,23 @@ zfs_rmnode(znode_t *zp)
 	}
 
 	/*
-	 * Free up all the data in the file.
+	 * Free up all the data in the file.  We don't do this for directories
+	 * because we need truncate and remove to be in the same tx, like in
+	 * zfs_znode_delete(). Otherwise, if we crash here we'll end up with
+	 * an inconsistent truncated zap object in the delete queue.  Note a
```

Comment:
@behlendorf I agree about breaking the guarantee, but I don't see anything that relies on that (i.e. that accesses the directory as a ZAP object after we start freeing it).  I'd like to at least understand the (performance) price of having the cleaner (never a broken ZAP) semantic.

I heard that you can end up with some large directories with Lustre -- do you know how many entries max?  According to my measurements, a directory would have about 6400 blocks (100 MB) per 1 million entries.  Which could take quite a while to read in, but *should* be OK since that should happen from dmu_tx_hold_free(), when we are not holding the transaction open (as long as we can cache all 800 KB of indirect blocks -- which seems fine).

Have you done any tests with deleting large directories to see if there's an impact on how long the transaction is assigned or how long it the spa_sync() takes to remove it?  From my calculations above I would expect it's fine with a directory that had 1,000,000 entries.